server/src/services/Google.ts

```diff
@@ -97,35 +114,37 @@ export async function storeGoogleTokens(code: string) {
   } else {
     const update = { access_token, expiry_date };
     // TODO: Handle the case where the refresh token is not sent *and* the
-    // record doesn't exist. We can make refresh_token nullable and that may be
-    // the best solution.
+    // record doesn't exist. If this happens, we need to redirect them to a
+    // new auth url, but with prompt: 'consent', so that Google will provide
+    // a new refresh token.
     await prisma.google_tokens.update({
       where: { id: TOKENS_ID },
       data: { ...update },
     });
   }
 }
 
-function createOAuth2Client() {
-  if (!keys) throw new Error('OAuth2 keys file missing');
-  return new OAuth2Client(
-    keys.client_id,
-    keys.client_secret,
-    keys.redirect_uris[0],
-  );
-}
-
-// TODO: use refresh tokens to get a new access token if the existing one is
-// expired or about to expire.
 async function createCredentialedClient() {
-  const oauth2Client = createOAuth2Client();
-  const { access_token } = await prisma.google_tokens.findFirstOrThrow();
-  oauth2Client.setCredentials({ access_token });
+  const oauth2Client = createOAuth2Client().on('tokens', onTokens);
+
+  const tokenInfo = await prisma.google_tokens.findFirstOrThrow();
+  const tokens = {
+    access_token: tokenInfo.access_token,
+    refresh_token: tokenInfo.refresh_token,
+    // Awkwardly, Prisma is setting the type to bigint, but it's really an 8 byte
+    // integer in the database.
+    expiry_date: tokenInfo.expiry_date as unknown as number,
```

Comment:
This might be weird question, but technically isn't that correct, since type in schema is changed to `BigInt`? Is schema change required?

Reply:
No, it's a good question.  Prisma's [BigInt](https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#bigint) corresponds to a Postgres [bigint](https://www.postgresql.org/docs/current/datatype-numeric.html) which is an 8 byte integer.  JavaScript's integers only go up to `2^53 -1` while Postgres can handle `2^63-1`.

...and the `number` coming back from Google is genuinely a JS number, since it's a unix timestamp (but in milliseconds). So, that easily fits in `2^53-1` hence Google giving it type `number`.

If that makes sense, I'll update the comment to explain this better.

Reply:
I somehow assumed timestamp will fit into the Postgres' _integer_. All is clear :+1:

Reply:
You and me both. It didn't occur to me until I realised that the fact everything was expiring in 1970 was because of integer overflow...